### DISCUSSION TITLE
create-next-app: Use shadcn/ui components for starter

### DISCUSSION THREAD

**Discussion by ctjlewis** [vercel/next.js#57787] [Ideas]
### Goals

- Reduce boilerplate in starter projects
- Handle things like automatic `min-height: 100vh` pages for the user out of the box

### Background

@rauchg described having to fiddle with viewport sizing in a recent project. I suggested patching the default CSS in #57231.

- The Geist `<Page>` component already [handles this automatically](https://geist-ui.dev/en-us/components/page). A similar component probably does/could exist in Shadcn's UI lib.

### Proposal

Why not rework the starter projects to use Shadcn UI by default? We could reduce the amount of junk code people start with and save them time on basic table-setting. 

It would be as easy, if not easier, to opt-out of this in the starter as it is to opt-out of today's starter code.

---

<sub>Originally thought the Geist UI library was Vercel's and proposed using that.</sub>

**Comment by leerob** [comment]
The biggest downside is more dependencies to install, and for folks to learn and understand. Granted, we're started to blur this line a bit by having Tailwind, but Shadcn UI takes this much further with a very opinionated set of tools.

**Reply by ctjlewis** [reply]
I was early with the TW feature in #24709 and I'm early here! To your point though, when Balasz eventually shipped that, it did cause a ton of commotion.

My thoughts:

  1. Vercel is in the business of setting de facto standards - first with SSR, now with Tailwind/TS opt-in, which I think has been vindicated by the largest product in modern history being built on the Next + TS + TW stack. It's uncomfortable to set standards but it moves the space forward and it's why you guys are where you are now, when 3 years ago people were still debating whether SPAs were better. (Many of those people are Next converts now.)
  
  2. Ultimately at the end of the day we *are* talking just about the default starter project, which has never really been some flawless unopinionated thing. That JSX/TSX source has always been at a little past "just good enough" state, and for a while (6 mos?), the page height was set to `100%` (not min-height) so the starter was totally broken on mobile. So it's seen rougher times without much downside.
  
      IMO when it comes to the starter project specifically, it's okay to get opinionated. It should reflect how *Vercel* thinks the best implementation of a simple Next project looks.
  
  3. (Probably most important) **Everyone ends up unplugging the starter pages anyway!** One `yarn remove` is all we'd really be adding here in exchange for the upside of less code in the starters, promoting better UI, and giving devs Vercel-blessed components out of the box.

I think it's a no-brainer and if I showed a before/after everyone would sign off.

**Reply by ctjlewis** [reply]
With regard to dependencies and bundle size, I know from having worked with the templates that the difference is negligible. With all the bells and whistles, TS + TW + ESLint etc, the impact of the component lib would be a non-issue IMO. You could install it ~400 times before it weighed as much as TSC (88kB v 35MB).

**Comment by ctjlewis** [comment]
One blocking issue is Shadcn's "no library" approach means we'd actually be *adding* source to templates, and not removing it...

> **What do you mean by not a component library?**
> 
> I mean you do not install it as a dependency. It is not available or distributed via npm.
> 
> Pick the components you need. Copy and paste the code into your project and customize to your needs. The code is yours.

https://ui.shadcn.com/docs
